From 64072679aae360299d92fe77f0fbe7140a92c5cb Mon Sep 17 00:00:00 2001
From: "Nguyen Van Linh [FGA.BU13]" <LinhNV25@fsoft.com.vn>
Date: Tue, 10 Apr 2018 08:29:40 +0700
Subject: [PATCH 272/628] spi: sh-msiof: Fix calculation of the division value

    commit <4c2d27539c63c40c748caeaae56de8abc84c5802>
    from git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-backport.git

    This patch is corresponding to the prohibitions of the hardware manual
    for R-Car Gen2 SoCs.
    In case of Gen2, BRDV[2:0]=B'111 is valid only
    when the BRPS[4:0] bits are set to B'00000 or B'00001.

Signed-off-by: Hiromitsu Yamasaki <hiromitsu.yamasaki.ym@renesas.com>
Signed-off-by: khanhluu <khanh.luu.xw@rvc.renesas.com>
Signed-off-by: Nguyen Van Linh [FGA.BU13] <LinhNV25@fsoft.com.vn>
Signed-off-by: vietn <vietn@fsoft.com.vn>
---
 drivers/spi/spi-sh-msiof.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/drivers/spi/spi-sh-msiof.c b/drivers/spi/spi-sh-msiof.c
index fa0e9f6..abd3ad3 100644
--- a/drivers/spi/spi-sh-msiof.c
+++ b/drivers/spi/spi-sh-msiof.c
@@ -180,6 +180,12 @@ struct sh_msiof_spi_priv {
 #define IER_RFUDFE	0x00000010 /* Receive FIFO Underflow Enable */
 #define IER_RFOVFE	0x00000008 /* Receive FIFO Overflow Enable */
 
+static int msiof_rcar_is_gen2(struct device *dev)
+{
+	struct device_node *node = dev->of_node;
+
+	return of_device_is_compatible(node, "renesas,msiof-r8a77470");
+}
 
 static u32 sh_msiof_read(struct sh_msiof_spi_priv *p, int reg_offs)
 {
@@ -272,6 +278,17 @@ static void sh_msiof_spi_set_clk_regs(struct sh_msiof_spi_priv *p,
 
 	k = min_t(int, k, ARRAY_SIZE(sh_msiof_spi_div_table) - 1);
 
+	/*
+	 * In case of Gen2, BRDV[2:0]=B'111 is valid only
+	 * when the BRPS[4:0] bits are set to B'00000 or B'00001.
+	 */
+	if (msiof_rcar_is_gen2(&p->pdev->dev) &&
+		sh_msiof_spi_div_table[k].brdv == SCR_BRDV_DIV_1 &&
+		!(brps == 1 || brps == 2)) {
+		k = 1; /* SCR_BRDV_DIV_1 -> SCR_BRDV_DIV_2 */
+		brps = DIV_ROUND_UP(brps, 2);
+	}
+
 	scr = sh_msiof_spi_div_table[k].brdv | SCR_BRPS(brps);
 	sh_msiof_write(p, TSCR, scr);
 	if (!(p->chipdata->master_flags & SPI_MASTER_MUST_TX))
-- 
2.7.4

